vsintegration/src/FSharp.Editor/LanguageService/FSharpProjectOptionsManager.fs

```diff
@@ -334,93 +334,150 @@ type private FSharpProjectOptionsReactor (workspace: Workspace, settings: Editor
                     return Some(parsingOptions, projectOptions)
         }
 
+    let tryGetOptionsByDocument (document: Document, ct: CancellationToken, userOpName) =
+        async {
+            if ct.IsCancellationRequested then
+                return None
+            else
+                try
+                    // For now, disallow miscellaneous workspace since we are using the hacky F# miscellaneous files project.
+                    if document.Project.Solution.Workspace.Kind = WorkspaceKind.MiscellaneousFiles then
+                        return None
+                    elif document.Project.IsFSharpMiscellaneousOrMetadata then
+                        let! options = tryComputeOptionsByFile document ct userOpName
+                        if ct.IsCancellationRequested then
+                            return None
+                        else
+                            return options
+                    else
+                        // We only care about the latest project in the workspace's solution.
+                        // We do this to prevent any possible cache thrashing in FCS.
+                        let project = document.Project.Solution.Workspace.CurrentSolution.GetProject(document.Project.Id)
+                        if not (isNull project) then
+                            let! options = tryComputeOptions project ct
+                            if ct.IsCancellationRequested then
+                                return None
+                            else
+                                return options
+                        else
+                            return None
+                with
+                | _ ->
+                    return None
+        }
+
+    let tryGetOptionsByProject (project: Project, ct: CancellationToken) =
```

Comment:
This is a lot of work to do ... holding a global lock.  Also ... the locking behaviour being different under release and debug is going to make debugging locking issues ... really tricky .... not that they aren't tricky already.

Reply:
The work isn't that much. It's not trying to type-check anything. At most, it may have to parse some part of the file to get the `#r ` references for scripts.

Reply:
@tihan question: is there a big implication if some areas that aren't too related to the core of the components using those abstractions, were to switch to the "RELEASE" behaviour, while keeping the core set enabled?

Are there roadblocks to do that if someone needs to work around how it works in debug session for any reason?

Reply:
> The work isn't that much. It's not trying to type-check anything. At most, it may have to parse some part of the file to get the #r references for scripts.

Will this also resolve and download nuget packages for `#r "nuget: ..."`?

Reply:
Good point, depends if `GetProjectOptionsFromScript` actually downloads/resolves nuget packages. If so, perhaps we could remove the lock for script files only.

Reply:
> Good point, depends if `GetProjectOptionsFromScript` actually downloads/resolves nuget packages. If so, perhaps we could remove the lock for script files only.

I think it will, it will call upon dependency manager provider's ResolveDependencies when building the LoadClosure.